diff --git a/main/apk-tools/APKBUILD b/main/apk-tools/APKBUILD
index ae2cac6..6b5dc7a 100644
--- a/main/apk-tools/APKBUILD
+++ b/main/apk-tools/APKBUILD
@@ -13,6 +13,7 @@ if [ "$CBUILD" = "$CHOST" ]; then
 	makedepends="$makedepends lua5.2-dev"
 fi
 source="http://dev.alpinelinux.org/archive/$pkgname/$pkgname-$pkgver.tar.xz
+	add-s390x.patch
 	"
 
 url="http://git.alpinelinux.org/cgit/apk-tools/"
@@ -82,6 +83,9 @@ luaapk() {
 	mv "$pkgdir"/usr/lib "$subpkgdir"/usr/lib/
 }
 
-md5sums="cc3efedf3047c256f13549ba58df3291  apk-tools-2.6.8.tar.xz"
-sha256sums="83f5cf9f0e75c19004a67150819eae2d4c859236b6c9c04d34d5b6968925f0d3  apk-tools-2.6.8.tar.xz"
-sha512sums="03fb42616335de9edaf3e44fd64a8d1c5dd8b0f25b7f3b176940b751a6547b8be153549f0cee9a5d65491025f6747799b6f5ba22b15d850593d499c07b5cc4c6  apk-tools-2.6.8.tar.xz"
+md5sums="cc3efedf3047c256f13549ba58df3291  apk-tools-2.6.8.tar.xz
+6bce9982a0b33f64d8fd14bcfbe1ec8c  add-s390x.patch"
+sha256sums="83f5cf9f0e75c19004a67150819eae2d4c859236b6c9c04d34d5b6968925f0d3  apk-tools-2.6.8.tar.xz
+c2b656c6a186f92276d45b9f47183dc2d3f0ec2229a9f6db561b6100544ac95e  add-s390x.patch"
+sha512sums="03fb42616335de9edaf3e44fd64a8d1c5dd8b0f25b7f3b176940b751a6547b8be153549f0cee9a5d65491025f6747799b6f5ba22b15d850593d499c07b5cc4c6  apk-tools-2.6.8.tar.xz
+33b80a4914c7f9915a2c890ddc72d56e2f7c0562f94a599f8a5b2365c97bdc7ddb07106341c93a98b555ace5f6a1a58365bdb2584c8f6e95ebe630771876d7cf  add-s390x.patch"
diff --git a/main/apk-tools/add-s390x.patch b/main/apk-tools/add-s390x.patch
new file mode 100644
index 0000000..e8a5363
--- /dev/null
+++ b/main/apk-tools/add-s390x.patch
@@ -0,0 +1,11 @@
+--- a/src/apk_defines.h
++++ b/src/apk_defines.h
+@@ -95,6 +95,8 @@ extern char **apk_argv;
+ #define APK_DEFAULT_ARCH	"armel"
+ #elif defined(__aarch64__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+ #define APK_DEFAULT_ARCH	"aarch64"
++#elif defined(__s390x__)
++#define APK_DEFAULT_ARCH	"s390x"
+ #else
+ #error APK_DEFAULT_ARCH not detected for this architecture
+ #endif
