diff --git a/main/musl/APKBUILD b/main/musl/APKBUILD
index 5f45210..0090c33 100644
--- a/main/musl/APKBUILD
+++ b/main/musl/APKBUILD
@@ -1,5 +1,6 @@
 # Contributor: William Pitcock <nenolod@dereferenced.org>
 # Maintainer: Timo Teräs <timo.teras@iki.fi>
+set -x
 pkgname=musl
 pkgver=1.1.15
 pkgrel=2
@@ -12,7 +13,7 @@ depends_dev="!uclibc-dev"
 makedepends="$depends_dev"
 subpackages="$pkgname-dev $pkgname-dbg libc6-compat:compat"
 [ "$BOOTSTRAP" != "nolibc" ] && subpackages="$subpackages $pkgname-utils"
-source="http://www.musl-libc.org/releases/musl-$pkgver.tar.gz
+source="https://github.com/tmh1999/alpine-bootstrap-s390x/raw/master/musl-1.1.15.tar.gz
 	0001-fix-asctime-day-month-names-not-to-vary-by-locale.patch
 
 	ldconfig
@@ -33,8 +34,10 @@ install_sysroot_headers() {
 	x86) ARCH="i386" ;;
 	x86_64) ARCH="x86_64" ;;
 	mips*) ARCH="mips" ;;
+	s390x) ARCH="s390x" ;;
 	esac
 	make ARCH="$ARCH" prefix=/usr DESTDIR="${CBUILDROOT}" install-headers || return 1
+
 }
 
 build() {
@@ -111,6 +114,7 @@ compat() {
 	x86) _ld="ld-linux.so.2" ;;
 	x86_64) _ld="ld-linux-x86-64.so.2" ;;
 	mips*) _ld="ld.so.1" ;;
+	s390*) _ld="ld-musl-s390x.so.1" ;;
 	esac
 
 	mkdir -p "$subpkgdir"/lib
@@ -121,21 +125,21 @@ compat() {
 	done
 }
 
-md5sums="9590a9d47ee64f220b3c12f7afb864ca  musl-1.1.15.tar.gz
+md5sums="cbeff14372a1f30e7b155337b2170547  musl-1.1.15.tar.gz
 c08825383e41e5dbcd3ffdfd2062dd47  0001-fix-asctime-day-month-names-not-to-vary-by-locale.patch
 830d01f7821b978df770b06db3790921  ldconfig
 0df687757221bbb0fc1aa67f1bd646f9  __stack_chk_fail_local.c
 57ef2c63b9ec6a2041694ace97d4ffa2  getconf.c
 2b941c4251cac44988a4abfc50e21267  getent.c
 45f92f8d59cf84d765de698a9578dbf4  iconv.c"
-sha256sums="97e447c7ee2a7f613186ec54a93054fe15469fe34d7d323080f7ef38f5ecb0fa  musl-1.1.15.tar.gz
+sha256sums="148af50993aa1d9516d05d0f3e7369420bd96288a730f45ef1e774c73877fb54  musl-1.1.15.tar.gz
 d157100aeed5b0866eb6d50288f63f26ea9900f1d4c7b8a1492294c912b5cc19  0001-fix-asctime-day-month-names-not-to-vary-by-locale.patch
 b4a2c06db38742e8c42c3c9838b285a7d8cdac6c091ff3df5ff9a15f1e41b9c7  ldconfig
 299a7d75a09de3e2e11e7fb4acc3182e4a14e868093d2f30938fce9bfcff13da  __stack_chk_fail_local.c
 d87d0cbb3690ae2c5d8cc218349fd8278b93855dd625deaf7ae50e320aad247c  getconf.c
 68373a55e89ce85c562d941ccf588337d6cc6c9c17689d695f65cd7607134bbe  getent.c
 f79a2930a2e5bb0624321589edf8b889d1e9b603e01e6b7ae214616605b3fdd7  iconv.c"
-sha512sums="9e923572c0d6bad3dc2d2646d4b0699c10b477ce6300ac6c6224895192a90667f581ddf6eda2ab8c4c16f47bde4bccb03bb90478638d136d9df721430f4d0163  musl-1.1.15.tar.gz
+sha512sums="0a0808e722d689d3275cb5db257917968501a084d57028776b1640fdf61021d17ecbfb73ffd282d7afa1cfa2ec3803ece383beb2704df332705b29659285c73c  musl-1.1.15.tar.gz
 1a74d5f5e0f6f2fe6029ed0f18b4603f80c990f19aa13d83c5d1f40f032b2ffb3819aae13ae1f96415bb08571774eec164e71d09028f2a5db4ae9b77e48cafe7  0001-fix-asctime-day-month-names-not-to-vary-by-locale.patch
 8d3a2d5315fc56fee7da9abb8b89bb38c6046c33d154c10d168fb35bfde6b0cf9f13042a3bceee34daf091bc409d699223735dcf19f382eeee1f6be34154f26f  ldconfig
 062bb49fa54839010acd4af113e20f7263dde1c8a2ca359b5fb2661ef9ed9d84a0f7c3bc10c25dcfa10bb3c5a4874588dff636ac43d5dbb3d748d75400756d0b  __stack_chk_fail_local.c
